prow/tide/tide.go

```diff
@@ -932,12 +932,54 @@ func (c *Controller) accumulateBatch(sp subpool) (successBatch []PullRequest, pe
 	return successBatch, pendingBatch
 }
 
+// prowJobsFromContexts constructs ProwJob objects from all successful presubmit contexts that include a baseSHA.
+// This is needed because otherwise we would always need retesting for results that are older than sinkers
+// max_prowjob_age.
+func prowJobsFromContexts(l *logrus.Entry, ghc githubClient, pr *PullRequest, baseSHA string) ([]prowapi.ProwJob, error) {
+	headContexts, err := headContexts(l, ghc, pr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get head contexts: %w", err)
+	}
+	var passingCurrentContexts []string
+	for _, headContext := range headContexts {
+		if headContext.State != githubql.StatusStateSuccess {
+			continue
+		}
+		if baseSHAForContext := config.BaseSHAFromContextDescription(string(headContext.Description)); baseSHAForContext != "" && baseSHAForContext == baseSHA {
+			passingCurrentContexts = append(passingCurrentContexts, string((headContext.Context)))
+		}
+	}
+
+	var prowjobsFromContexts []prowapi.ProwJob
+	for _, passingCurrentContext := range passingCurrentContexts {
+		prowjobsFromContexts = append(prowjobsFromContexts, prowapi.ProwJob{
+			Spec: prowapi.ProwJobSpec{
+				Context: passingCurrentContext,
+				Refs:    &prowapi.Refs{Pulls: []prowapi.Pull{{Number: int(pr.Number), SHA: string(pr.HeadRefOID)}}},
```

Comment:
This is a bit confusing - instead of only creating ProwJobs that pass against the current HEAD (I assume we append this to the filtered list?) why don't we just always create the synthetic ProwJob and have it specify the base ref it tested against, then have the filtering etc happen as normal after?

Reply:
Because the list this gets appended to is already filtered by baseSHA

Reply:
Hm - I guess this entire change would be less dangerous etc if we always created fake ProwJobs with whatever baseSHA and status we see in GitHub, then we'd only have one filtering code for the list and not potentially introduce two areas that, if bugged, have large impacts on tide.

Reply:
Updated it to create ProwJobs instead. It will still only create prowjobs for the matching baseSHA and successes to not overload the apiserver, but it will take the created baseSHA from the github context.

Reply:
Why do the ProwJobs need to be added to the API server? That seems like unnecessary load on the API server given that Tide is the only thing that will consume these. There are also other catches like avoiding creating the same PJs every sync loop and ensuring other Prow components handle the fake PJs appropriately (e.g. Deck should ignore them, sinker would need to handle them differently).
I interpreted Steve's request to mean create ProwJob structs so that we can filter them the same way we filter real PJs, not create real ProwJob CRs on the API server.

Reply:
> Why do the ProwJobs need to be added to the API server?

Because the filtering is done via [an index over the cache](https://github.com/kubernetes/test-infra/blob/eeb7afb52810b459ddc23b33890e69cbf8efb61f/prow/tide/tide.go#L1716) in order to be efficient and I don't think trying to trick the cache into thinking there was data that isn't on the server is a great idea.

> There are also other catches like avoiding creating the same PJs every sync loop and ensuring other Prow components handle the fake PJs appropriately 

True, I've reverted the pull to the original version

> I interpreted Steve's request to mean create ProwJob structs so that we can filter them the same way we filter real PJs, not create real ProwJob CRs on the API server.

It already did create prowjob structs in the first version and as mentioned above, re-using the filtering isn't really possible.